test/model/typing_status_test.dart

```diff
@@ -210,4 +248,247 @@ void main() {
+
+    Future<void> prepare() async {
+      addTearDown(testBinding.reset);
+      // Only test with the latest behavior of the setTypingStatus API.
+      // (Legacy variations are covered by setTypingStatus's own tests.)
+      final account = eg.account(
+        user: eg.selfUser, zulipFeatureLevel: eg.futureZulipFeatureLevel);
+      store = eg.store(account: account, initialSnapshot: eg.initialSnapshot(
+        zulipFeatureLevel: eg.futureZulipFeatureLevel));
+      model = store.typingNotifier;
+      connection = store.connection as FakeApiConnection;
+
+      final channel = eg.stream();
+      await store.addStream(channel);
+      await store.addSubscription(eg.subscription(channel));
+      narrow = TopicNarrow(channel.streamId, 'topic');
+    }
+
+    /// Prepares store and triggers a "typing started" notice.
+    Future<void> prepareStartTyping(FakeAsync async) async {
+      await prepare();
+      connection.prepare(json: {});
+      await model.keystroke(narrow);
+      checkTypingRequest(TypingOp.start, narrow);
+      check(async.pendingTimers).single;
+    }
+
+    test('start typing repeatedly extends idle timer', () => awaitFakeAsync((async) async {
+      // t = 0ms: Start typing. The idle timer is set to typingStoppedWaitPeriod.
+      await prepareStartTyping(async);
+
+      const waitTime = Duration(milliseconds: 100);
+      // [waitTime] should not be long enough
+      // to trigger a "typing stopped" notice.
+      assert(waitTime < model.typingStoppedWaitPeriod);
+
+      async.elapse(waitTime);
+      // t = 100ms: The idle timer is reset to typingStoppedWaitPeriod.
+      connection.prepare(json: {});
+      await model.keystroke(narrow);
+      check(connection.lastRequest).isNull();
+      check(async.pendingTimers).single;
+
+      async.elapse(model.typingStoppedWaitPeriod - const Duration(milliseconds: 1));
+      // t = typingStoppedWaitPeriod + 99ms:
+      //   Since the timer was reset at t = 100ms, the "typing stopped" notice has
+      //   not been sent yet.
+      check(connection.lastRequest).isNull();
+      check(async.pendingTimers).single;
+
+      async.elapse(waitTime);
+      // t = typingStoppedWaitPeriod + 100ms:
+      //   The new timer expires and the "typing stopped" notice is sent.
+      checkTypingRequest(TypingOp.stop, narrow);
+    }));
+
+    test('start typing repeatedly does not resend "typing started" notices', () => awaitFakeAsync((async) async {
+      // t = 0ms: Start typing.
+      await prepareStartTyping(async);
+
+      const waitInterval = Duration(milliseconds: 2000);
+      // [waitInterval] should not be long enough
+      // to trigger a "typing stopped" notice.
+      assert(waitInterval < model.typingStoppedWaitPeriod);
+      // [waitInterval] should not be long enough either,
+      // that we only send the "typing started" notice once.
+      assert(waitInterval < model.typingStartedWaitPeriod);
```

Comment:
```suggestion
      // [waitInterval] should be short enough
      // that the loop below runs more than once.
      assert(waitInterval < model.typingStartedWaitPeriod);
```

I read 'send the "typing started" notice' as meaning actually making the request, not just calling `model.keystroke`.